### PR TITLE
New checkout

### PR DESCRIPTION
**Pull request by mitcom**
Fixes: #379

### CODE REVIEW

File: saleor/checkout/core.py [hunk 1]
```diff
@@ -46,85 +46,46 @@ def __iter__(self):
         return iter(self.steps)
 
     def generate_steps(self, cart):
+        shipping_address = None
```
Comment (patrys) [inline saleor/checkout/core.py]: What is the purpose of this?

File: saleor/checkout/core.py [hunk 2]
```diff
@@ -46,85 +46,46 @@ def __iter__(self):
...
-        self.billing = BillingAddressStep(
-            self.request, self.get_storage('billing'))
-        self.steps.append(self.billing)
+
         if self.is_shipping_required():
-            self.shipping = ShippingStep(
-                self.request, self.get_storage('shipping'),
-                self.cart, billing_address=self.billing_address)
+            self.shipping = ShippingAddressStep(
+                self.request, self.storage['shipping_address'], checkout=self)
```
Comment (patrys) [inline saleor/checkout/core.py]: What happened to `self.get_storage(…)`?

File: saleor/checkout/steps.py [hunk 3]
```diff
@@ -65,156 +141,132 @@ def validate(self):
...
 
-    def process(self, extra_context=None):
-        context = dict(extra_context or {})
-        context['delivery_form'] = self.forms['delivery']
-        return super(ShippingStep, self).process(extra_context=context)
+    def validate(self):
+        selected_method_name = self.storage.get('delivery_method')
+        valid_methods = [d['method'].name for d in self.available_deliveries]
+        if selected_method_name not in valid_methods:
+            raise InvalidData('Select a valid delivery method')
```
Comment (bogdal) [inline saleor/checkout/steps.py]: Please use `ugettext_lazy`. We should be able to translate this.

File: saleor/checkout/core.py [hunk 4]
```diff
@@ -46,85 +46,46 @@ def __iter__(self):
...
+            shipping_address = self.shipping.address
             self.steps.append(self.shipping)
+            self.delivery = ShippingMethodStep(
```
Comment (patrys) [inline saleor/checkout/core.py]: I don't like having `self.shipping` and `self.delivery` as both seem to mean the same thing. Explicitly call them `shipping_address` and `shipping_method`.

File: saleor/checkout/forms.py [hunk 5]
```diff
@@ -1,18 +1,42 @@
...
+            return super(AddressChoiceField, self).validate(value)
 
-    shipping_same_as_billing = forms.BooleanField(initial=True, required=False)
+    def to_python(self, value):
+        if value in self.choice_values:
+            return value
+        else:
+            return super(AddressChoiceField, self).to_python(value)
 
+    def add_magic_choices(self, magic_choices, queryset):
```
Comment (patrys) [inline saleor/checkout/forms.py]: This seems to cross layers of separation. Can't we pass all necessary information using `__init__` and `__iter__`?

File: saleor/checkout/forms.py [hunk 6]
```diff
@@ -1,18 +1,42 @@
...
+    address = AddressChoiceField(
+        queryset=Address.objects.none(),
+        widget=forms.RadioSelect)
+
+    def __init__(self, queryset, possibilities, *args, **kwargs):
+        super(UserAddressesForm, self).__init__(*args, **kwargs)
+        address_field = self.fields['address']
+        address_field.queryset = queryset if queryset else Address.objects.none()
+        address_field.add_magic_choices(possibilities, queryset)
+        address_field.choice_values = [value for value, label in possibilities]
```
Comment (patrys) [inline saleor/checkout/forms.py]: Have the field initialize itself. The class should encapsulate its logic, not depend on an external entity settings its internals.

**Pull request closed without merging**